lib/checkstl.cpp

```diff
@@ -148,8 +158,9 @@ void CheckStl::outOfBounds()
                     continue;
                 if (!value.errorSeverity() && !mSettings->severity.isEnabled(Severity::warning))
                     continue;
-                if (value.intvalue == 0 && (indexTok || (containerYieldsElement(container, parent) &&
-                                                         !containerAppendsElement(container, parent)))) {
+                if (value.intvalue == 0 && (indexTok ||
+                                            (containerYieldsElement(container, parent) && !containerAppendsElement(container, parent)) ||
+                                            (value.isKnown() && containerPopsElement(container, parent)))) {
```

Comment:
We should be able to warn when the value is not known as well.

Reply:
The problem is code like
~~~
void f(std::vector<int>& v) {
	v.pop_back();
	if (v.empty()) {}
}
~~~
We don't seem to take into account that the size changes with `pop_back()`.

Reply:
That looks like a bug in Valueflow. Maybe we don't account for the POP action.

Reply:
It's not necessarily a ValueFlow issue.
We don't warn for this either, although `v.resize()` does get a possible empty value:
~~~
void g(std::vector<int>& v) {
	int n = v.size();
	v.resize(n - 1);
	if (v.empty()) {}
}
~~~
`containerYieldsElement()` is obviously false, and `containerAppendsElement()` does not check whether `resize()` adds or removes elements.

Reply:
We also don't warn for
~~~
void g(std::vector<int>& v) {
	v.push_back(1);
	if (v.empty()) {}
}
~~~

Reply:
So I fixed the valueflow issues in #4800 and #4803 so you should be able to warn for possible values as well. 

For the `v.resize(n - 1)` case, the checker would need to be updated, but I dont think it needs done in this PR though.

Reply:
Nice. So we should be able to warn for
~~~
void f(std::vector<int>& v) {
	if (v.empty()) {}
	v.pop_back();
}
~~~
after those PRs have been merged.
The `resize(n - 1)` example is equivalent to the `pop_back()` one in the 2nd comment above, so we shouldn't warn for those.

Reply:
I think #4800 is only needed to support possible values as #4803 needs some more work to fix FPs.